### PR TITLE
kvserver: fully adopt useReproposalsV2

### LINKED ISSUE #105625: kvserver: phase out useReproposalsV2
This is introduced in https://github.com/cockroachdb/cockroach/pull/97779 to keep both the old and new way of reproposing with a new lease applied index around.

In a few weeks, we should remove the old way, i.e. hard-code `useReproposalsV2` to `true` and simplify code based on that.

Epic: CRDB-25287

Jira issue: CRDB-29130

### PR DESCRIPTION
**Pull request by tbg**
This hard-codes useReproposalsV2 to true (prior to this PR, it's metamorphic with a default of true) and applies the resulting simplifications across the codebase.

These include simplified control flow and an overall simplified architecture.

Closes https://github.com/cockroachdb/cockroach/issues/105625.

Epic: CRDB-25287
Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_application_decoder.go [hunk 1]
```diff
@@ -75,15 +75,13 @@ func (d *replicaDecoder) decode(ctx context.Context, ents []raftpb.Entry) error
...
-// complexity of having multiple entries associated to the same proposal during
-// application.
-func (d *replicaDecoder) retrieveLocalProposalsV2() (anyLocal bool) {
+// retrieveLocalProposals removes all proposals which have a log entry pending
+// immediate application from the proposals map. The entries which are paired up
+// with a proposal in that way are considered "local", meaning a client is
+// waiting on their result, and may be reproposed (as a new proposal) with a new
+// lease index in case they apply with an illegal lease index (see
+// tryReproposeWithNewLeaseIndex).
+func (d *replicaDecoder) retrieveLocalProposals() (anyLocal bool) {
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_application_decoder.go:84]: nit: V2 rename should be in previous commit, or alternatively just squash the entire commit into it.

File: pkg/kv/kvserver/replica_proposal.go [hunk 2]
```diff
@@ -52,69 +52,59 @@ import (
...
+//	         │         ▼       [4]
+//	         │    proposal map───►proposal map
+//	         │         │          and propBuf
+//	         │         │[3]            │
+//	         │         │               │
+//	         │         ▼               │[4]
+//	         └──apply goroutine◄───────┘
+//	                   │
+//	                   ▼
+//	                finished
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_proposal.go:73]: ❤️

**Review by erikgrinaker — APPROVED** [review]
LGTM, thanks for the cleanup!

**Comment by tbg** [comment]
TFTR!

Comments addressed.

bors r=erikgrinaker

**Comment by tbg** [comment]
TestRules/join_order flaked, pointed it out [here](https://cockroachlabs.slack.com/archives/C016CAD2HQ8/p1689743545636259)

bors r=erikgrinaker

**Pull request merged**